### PR TITLE
[hail] Add CombOpValue and AggStateValue nodes for agg lowering

### PR DESCRIPTION
**Pull request by tpoterba**
stacked on #7882

### CODE REVIEW

File: hail/src/main/scala/is/hail/expr/ir/EmitFunctionBuilder.scala [hunk 1]
```diff
@@ -363,6 +363,12 @@ class EmitFunctionBuilder[F >: Null](
     _hfield.load()
   }
 
+  val lazyFieldMemo: mutable.Map[Any, LazyFieldRef[_]] = mutable.Map.empty
+
+  def getOrDefineLazyField[T: TypeInfo](setup: Code[T], id: Any): LazyFieldRef[T] = {
```
Comment (patrick-schultz) [inline hail/src/main/scala/is/hail/expr/ir/EmitFunctionBuilder.scala]: Is there any reason this shouldn't be on `FunctionBuilder`?

File: hail/src/main/scala/is/hail/expr/ir/IR.scala [hunk 2]
```diff
@@ -340,6 +340,8 @@ final case class InitOp(i: Int, args: IndexedSeq[IR], aggSig: PhysicalAggSignatu
 final case class SeqOp(i: Int, args: IndexedSeq[IR], aggSig: PhysicalAggSignature) extends IR
 final case class CombOp(i1: Int, i2: Int, aggSig: PhysicalAggSignature) extends IR
 final case class ResultOp(startIdx: Int, aggSigs: IndexedSeq[PhysicalAggSignature]) extends IR
+final case class CombOpValue(i1: Int, value: IR, aggSig: PhysicalAggSignature) extends IR
```
Comment (patrick-schultz) [inline hail/src/main/scala/is/hail/expr/ir/IR.scala]: Just `i`?

File: hail/src/test/scala/is/hail/expr/ir/Aggregators2Suite.scala [hunk 3]
```diff
@@ -743,4 +743,31 @@ class Aggregators2Suite extends HailSuite {
...
+          InitOp(0, FastSeq(I32(5)), takeSig),
+          SeqOp(0, FastSeq(NA(TInt64())), takeSig),
+          SeqOp(0, FastSeq(I64(-1l)), takeSig),
+          SeqOp(0, FastSeq(I64(2l)), takeSig)
+        )),
+        AggStateValue(0, takeSig),
+        FastIndexedSeq(takeSig)),
+      RunAgg(
+        Begin(FastSeq(
+          InitOp(0, FastSeq(I32(10)), takeSig),
```
Comment (patrick-schultz) [inline hail/src/test/scala/is/hail/expr/ir/Aggregators2Suite.scala]: As discussed, we should probably require agg state values to be combined with agg states that were initialized with the same InitOp args.

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala [hunk 4]
```diff
@@ -969,6 +969,29 @@ private class Emit(
...
+        val t = value.pType.asInstanceOf[PBinary]
+        val xValue = emit(value)
+        EmitTriplet(
+          Code(xValue.setup,
+          xValue.m.mux(
+            Code._fatal("cannot combOp a missing value"),
+            Code(
+              newState.createState,
+              newState.deserializeFromBytes(t, coerce[Long](xValue.v)),
+              rvAgg.combOp(sc.states(i), newState)
```
Comment (patrick-schultz) [inline hail/src/main/scala/is/hail/expr/ir/Emit.scala]: I don't remember how agg states work well enough. Does anything need to be done to free `newState`?

File: hail/src/main/scala/is/hail/expr/ir/EmitFunctionBuilder.scala (hunk 1 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/EmitFunctionBuilder.scala]: er, that would be fine too.

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala (hunk 4 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/Emit.scala]: It'll get cleaned up in RunAgg, I believe.

**Pull request merged**